include/envoy/http/original_ip_detection.h

```diff
@@ -0,0 +1,55 @@
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "envoy/common/pure.h"
+#include "envoy/config/typed_config.h"
+#include "envoy/http/header_map.h"
+#include "envoy/network/address.h"
+
+namespace Envoy {
+namespace Http {
+
+struct OriginalIPDetectionParams {
+  Http::RequestHeaderMap& request_headers;
+};
+
+/**
+ * Interface class for original IP detection extensions.
+ */
+class OriginalIPDetection {
+public:
+  virtual ~OriginalIPDetection() = default;
+
+  /**
+   * Detect the final remote address.
+   *
+   * @param param supplies the OriginalIPDetectionParams params for detection.
```

Comment:
@return comments?

Reply:
Hmm, that's an interesting thought. In our case, I was planning to set a header when the detection failed (e.g.: `x-pinterest-detection-failed`) and then reject it further down from a companion filter. 

But being able to reject from the extension itself would indeed free us from the need of an additional filter.

Reply:
I think per Mark's preference we'd not want failover logic, but instead your extension could check for "nothing returned" and call the "base class"/"default extension" WDYT?